### PR TITLE
Transition to google.golang.org/protobuf

### PR DESCRIPTION
**Pull request by nkuba**
As the `github.com/gogo/protobuf` package got deprecated we move to `google.golang.org/protobuf`. In this PR we do most of the transition.

### Generate `%.pb.go` files from `%.proto` (186bd8d7f83ea5e15a6fa2e08e1bf0c7965c1de3)

We don't need to maintain multiple `gen.go` files for protobuf
generation. We defined a makefile's target for protobuf generation in the
main `Makefile`.

To run it execute `make gen_proto` in the project root.

We added a `go:generate` entry to the `main.go` so if someone is used to
using `go generate .` instead of `make` in our codebase instead of `make`
the new makefile's target will be executed.

Makefile gives an advantage of regenerating the code only if the
prerequisites have changed. As prerequisites, we defined `%.proto` files,
and also `go.mod`, `go.sum` (in case a dependency version is updated). The
code will be regenerated only if once of the files is updated.

We include regenerated `%.pb.go` files in this PR. (86762731e166eeefa3ee7175141c69967098f18a)

### Marshaling/Unmarshaling (51c1d17011c319d0ab2a20515661256e457e4aa3)

Structs generated for protobug no longer implement self-marshaling/self-unmarshaling
functions from `proto.Marshaler` and `proto.Unmarshaler` interfaces.

We switch to the recommended marshaling and unmarshaling functions from
google.golang.org/protobuf/proto: `proto.Marshal` and `proto.Unmarshal`.

### Update `libp2p.authenticatedConnection` (b789dc463512bf6e0f58252e43bdb8def732eb68)

Unfortunately, the new package doesn't expose `NewDelimitedReader` and
`NewDelimitedWriter`, so we had to update the code.

We were lucky that `google.golang.org/protobuf/proto` has support for
delimited read and writer already implemented, but this has not been
released yet, so we need to pin to a commit from the master branch. We do
that only for the `protodelim` feature. The rest of the imports will remain
on the released version of google.golang.org/protobuf/proto (v.1.28.1)

### gogo/protobuf remainings

In a follow-up PRs we should handle remaining references to the `gogo/protobuf`:
- `net` package
  https://github.com/keep-network/keep-core/blob/4425e32d77e50b31da0d1e7d85f1de91dd77b91a/pkg/net/net.go#L32-L34
  https://github.com/keep-network/keep-core/blob/4425e32d77e50b31da0d1e7d85f1de91dd77b91a/pkg/net/net.go#L85-L87
- `pbutils` 
  https://github.com/keep-network/keep-core/blob/4425e32d77e50b31da0d1e7d85f1de91dd77b91a/pkg/internal/pbutils/pbutils.go#L24-L30

### CODE REVIEW

File: pkg/net/libp2p/authenticated_connection.go [hunk 1]
```diff
@@ -212,24 +249,19 @@ func (ac *authenticatedConnection) initiatorSendAct1(
...
-func (ac *authenticatedConnection) initiatorReceiveAct2(
-	initiatorConnectionReader protoio.ReadCloser,
-) (*handshake.Act2Message, error) {
+func (ac *authenticatedConnection) initiatorReceiveAct2() (*handshake.Act2Message, error) {
 	var (
 		act2Envelope pb.HandshakeEnvelope
 		act2Message  = &handshake.Act2Message{}
 	)
-	if err := initiatorConnectionReader.ReadMsg(&act2Envelope); err != nil {
+	fmt.Printf("initiator waiting to receive message")
```
Comment (pdyraga) [inline pkg/net/libp2p/authenticated_connection.go:263]: 🔥

File: pkg/net/libp2p/authenticated_connection.go [hunk 2]
```diff
@@ -327,16 +349,16 @@ func (ac *authenticatedConnection) runHandshakeAsResponder() error {
...
-func (ac *authenticatedConnection) responderReceiveAct1(
-	responderConnectionReader protoio.ReadCloser,
-) (*handshake.Act1Message, error) {
+func (ac *authenticatedConnection) responderReceiveAct1() (*handshake.Act1Message, error) {
 	var (
 		act1Envelope pb.HandshakeEnvelope
 		act1Message  = &handshake.Act1Message{}
 	)
-	if err := responderConnectionReader.ReadMsg(&act1Envelope); err != nil {
+	fmt.Println("responder waiting to receive message")
```
[the remark this answers is not in the document]
Comment (pdyraga) [reply pkg/net/libp2p/authenticated_connection.go]: b1768f2684001c8fc817b562911bb511b23234a1

File: pkg/net/libp2p/authenticated_connection.go [hunk 3]
```diff
@@ -327,16 +349,16 @@ func (ac *authenticatedConnection) runHandshakeAsResponder() error {
...
+	if err := ac.pipe.receive(&act1Envelope); err != nil {
 		return nil, err
 	}
+	fmt.Println("got message")
```
[the remark this answers is not in the document]
Comment (pdyraga) [reply pkg/net/libp2p/authenticated_connection.go]: b1768f2684001c8fc817b562911bb511b23234a1

File: pkg/net/libp2p/authenticated_connection.go (hunk 1 above)
Comment (pdyraga) [reply pkg/net/libp2p/authenticated_connection.go]: b1768f2684001c8fc817b562911bb511b23234a1

**Comment by pdyraga** [comment]
I tested this PR locally with 2 clients working together. They were able to connect to each other, perform random beacon key generation 3 times, and ECDSA key generation 2 times. They were able to save and load random beacon and ECDSA key shares from the storage.

I have also tested the scenario when a modified client tries to connect to a bootstrap node with no stake - the handshake mechanism properly rejected the connection.

**Comment by pdyraga** [comment]
There are two problems reported by `go-vet` that were not caused by the code from this PR but that we should fix in one of the next PRs when working on the transition to google.golang.org/protobuf. They should be fairly easy to solve:
```
pkg/net/libp2p/channel.go:255:60: call of c.processContainerMessage copies lock value: github.com/keep-network/keep-core/pkg/net/gen/pb.BroadcastNetworkMessage contains google.golang.org/protobuf/internal/impl.MessageState contains sync.Mutex
pkg/net/libp2p/channel.go:260:10: processContainerMessage passes lock by value: github.com/keep-network/keep-core/pkg/net/gen/pb.BroadcastNetworkMessage contains google.golang.org/protobuf/internal/impl.MessageState contains sync.Mutex
```

**Review by pdyraga — APPROVED** [review]

File: go.mod [hunk 4]
```diff
@@ -7,6 +7,11 @@ go 1.18
 // https://github.com/bnb-chain/tss-lib/pull/203
 replace github.com/bnb-chain/tss-lib => github.com/keep-network/tss-lib v1.3.3-0.20220826121242-73ee9558285f
 
+replace (
```
[the remark this answers is not in the document]
Comment (nkuba) [reply go.mod:10]: https://github.com/keep-network/keep-core/pull/3263

File: docs-v1/development/README.adoc [hunk 5]
```diff
@@ -100,7 +100,7 @@ branches are instead published to `+https://docs.keep.network/<branch name>/+`.
 
 - Please avoid using `~` when defining `$GOBIN` location and use `$HOME` instead.
   We've been observing https://github.com/google/protobuf/issues/3355[some issues]
-  with locating `protoc-gen-gogoslick` when running `go generate` and `$GOBIN`
+  with locating `protoc-gen-go` when running `go generate` and `$GOBIN`
   contained `~`.
```
[the remark this answers is not in the document]
Comment (nkuba) [reply docs-v1/development/README.adoc:104]: This is a doc for v1, it shouldn't be updated in the first place :) Can we leave it as is?

**Pull request merged**